### PR TITLE
Refactored iOS Object Detector

### PR DESCRIPTION
**Pull request by priankakariatyml**
### Description
Re-architected iOS Object Detection sample.

Fixes # (issue)
1. Fixed unresponsive UI.
2. Rearchitected the code
       - Split camera and library handling into separate view controllers
       - Updates to how maxResults, scoreThreshold and model changes are handled.
       - Re-designed OverlayView drawing, ObjectDetectorHelper classes
       - Refactored code to more readable, shorter and fail safe wherever possible according to Swift standards.
4. Fixed drawing of bounding boxes

### Checklist

### CODE REVIEW

**Comment by PaulTR** [comment]
Hey Prianka! Would you mind pulling in main to fix the merge conflicts? Thanks!

File: examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    set {
+      objectDetectorServiceQueue.async(flags: .barrier) {
+        self._objectDetectorService = newValue
+      }
+    }
+  }
+  
+  override func viewWillAppear(_ animated: Bool) {
+    super.viewWillAppear(animated)
+    print("CameraController Appear")
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift:64]: nit: Remove this log as we don't print it for other view controllers.

File: examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+#if !targetEnvironment(simulator)
+    initializeObjectDetectorServiceOnSessionResumption()
+    cameraCapture.checkCameraConfigurationAndStartSession()
+#endif
+  }
+  
+  override func viewWillDisappear(_ animated: Bool) {
+    super.viewWillDisappear(animated)
+    print("CameraController Disappear")
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift:73]: ditto.

File: examples/object_detection/ios/ObjectDetector/ViewContollers/DetectorViewController.swift [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import MediaPipeTasksVision
+
+/**
+ * Base class to view controllers that handle performing detection on images, videos or camera frames.
+ */
+class DetectorViewController: UIViewController {
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/DetectorViewController.swift:21]: I think we should have less number of classes so that the code would become easier to read for developers who came to this sample.

Therefore, I don't think we should have this base class but fold the feature implemented here into `CameraViewController` and `MediaLibraryViewController` because:
* DetectorMetadata observer is only used in CameraViewController
* `draw` method doesn't do anything meaningful. It just pass a list of parameters to overlayView for drawing.

File: examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+  }
+  
+  private func clearObjectDetectorServiceOnSessionInterruption() {
+    removeDetectorMetadataKeyValueObservers()
+    objectDetectorService = nil
+  }
+}
+
+extension CameraViewController: CameraFeedManagerDelegate {
+  override func observeValue(
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift:128]: Move this method because it doesn't belong to the CameraFeedManagerDelegate protocol.

File: examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import UIKit
+
+import MediaPipeTasksVision
+
+// Helper Methods for calculating bounding boxes that are to be drawn.
+class ObjectOverlayHelper {
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift:21]: IMO having less class is better for sample apps. Can you move this helper logic into OverlayView so that all logic necessary to display the object detection result are in the same class?

File: examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+
+import UIKit
+import AVFoundation
+
+/**
+ Displays a preview of the image being processed. By default, this uses the device's camera frame,
+ but will use a still image copied from clipboard if `shouldUseClipboardImage` is set to true.
+ */
+class PreviewView: UIView {
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift:22]: This is only used in the CameraViewController, and the logic is quite simple. Rather than defining a custom class, can you just implement it in `CameraViewController` to reduce the number of classes?

File: examples/object_detection/ios/ObjectDetector.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -556,7 +612,7 @@
 					"@executable_path/Frameworks",
 				);
 				MARKETING_VERSION = 1.0;
-				PRODUCT_BUNDLE_IDENTIFIER = com.google.mediapipe.examples.objectdetection;
+				PRODUCT_BUNDLE_IDENTIFIER = com.prianka.ImageClassification;
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector.xcodeproj/project.pbxproj:615]: Please use the sample bundleId as the original app

File: examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+  
+  private var isSessionRunning = false
+  private let backgroundQueue = DispatchQueue(label: "com.cameraController.backgroundQueue")
+  
+  // MARK: Controllers that manage functionality
+  // Handles all the camera related functionality
+  private lazy var cameraCapture = CameraFeedManager(previewView: previewView)
+  
+  private let objectDetectorServiceQueue = DispatchQueue(
+    label: "objectDetectorServiceQueue",
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift:43]: Please prefix the queue name with `com.google.mediapipe`

File: examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+  func shouldClicksBeEnabled(_ isEnabled: Bool)
+}
+
+class CameraViewController: DetectorViewController {
+  @IBOutlet weak var previewView: PreviewView!
+  @IBOutlet weak var cameraUnavailableLabel: UILabel!
+  @IBOutlet weak var resumeButton: UIButton!
+  
+  private var isSessionRunning = false
+  private let backgroundQueue = DispatchQueue(label: "com.cameraController.backgroundQueue")
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/CameraViewController.swift:36]: Please prefix the queue name with `com.google.mediapipe`

File: examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+        break;
+    }
+  }
+  
+  private func playVideo(mediaURL: URL, resultBundle: ResultBundle?) {
+    playVideo(asset: AVAsset(url: mediaURL))
+    playerTimeObserverToken = playerViewController?.player?.addPeriodicTimeObserver(
+      forInterval: CMTime(value: Constants.inferenceTimeIntervalMs,
+                          timescale: Int32(Constants.kMilliSeconds)),
+      queue: DispatchQueue(label: "timeObserverQueue", qos: .userInteractive),
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift:291]: Please prefix the queue name with `com.google.mediapipe`

File: examples/object_detection/ios/ObjectDetector/Managers/CameraFeedManager.swift [hunk 11]
```diff
@@ -64,43 +64,46 @@ enum CameraConfiguration {
...
+          return CGSize(width: minDimension, height: maxDimension)
+      }
+    }
+  }
 
-  // MARK: Camera Related Instance Variables
+  // MARK: Instance Variables
   private let session: AVCaptureSession = AVCaptureSession()
   private let previewView: PreviewView
   private let sessionQueue = DispatchQueue(label: "sessionQueue")
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/Managers/CameraFeedManager.swift:92]: Please prefix the queue name with `com.google.mediapipe`

File: examples/object_detection/ios/ObjectDetector/Managers/ObjectDetectorService.swift [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+protocol ObjectDetectorServiceVideoDelegate: AnyObject {
+ func objectDetectorService(_ objectDetectorService: ObjectDetectorService,
+                                  didFinishDetectionOnVideoFrame index: Int)
+ func objectDetectorService(_ objectDetectorService: ObjectDetectorService,
+                             willBeginDetection totalframeCount: Int)
+}
+
+
+// Initializes and calls the MediaPipe APIs for detection.
+class ObjectDetectorService: NSObject {
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/Managers/ObjectDetectorService.swift:40]: Can it be `ObjectDetectorManager` to go along with `CameraFeedManager`? I notice that you put it into a folder called `Managers`.

File: examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift [hunk 13]
```diff
@@ -0,0 +1,358 @@
...
+    playerViewController?.player?.pause()
+    playerViewController?.player = nil
+  }
+  
+  private func removeObservers(player: AVPlayer?) {
+    guard let player = player else {
+      return
+    }
+    
+    if let timeObserverTokern = playerTimeObserverToken {
```
Comment (khanhlvg) [inline examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift:126]: typo: timeObserverToken

**Comment by khanhlvg** [comment]
Thanks Prianka for refactoring the app! A few more comments:
* Please set Team to None when you commit the code
image
* Please use the sample bundleId as the original app (``)

File: examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift (hunk 5 above)
Comment (priankakariatyml) [reply examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift:21]: The helper imports MediaPipeTasksVision. I did not want the view to import it. Just to separate any detection to overlay conversion logic that the view doesn't need to know.

But it is not a necessity. LMK if this needs to be done.

File: examples/object_detection/ios/ObjectDetector/Managers/ObjectDetectorService.swift (hunk 12 above)
Comment (priankakariatyml) [reply examples/object_detection/ios/ObjectDetector/Managers/ObjectDetectorService.swift:40]: I didn't want to create another folder. Hence put the service in there. Will it be more appropriate to rename CameraFeedManager to CameraFeedService? Since the app is not all about live camera anymore, service sounds better since it handles only functionality specific to one part of the app. Manager sounds more global.

Comment (khanhlvg) [reply examples/object_detection/ios/ObjectDetector/Managers/ObjectDetectorService.swift:40]: `CameraFeedService` sgtm

File: examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift (hunk 5 above)
Comment (khanhlvg) [reply examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift:21]: `OverlayView` is meant to draw an object returned by `MediaPipeTasksVision` so I don't see having it to import `MediaPipeTasksVision` being a problem. Can you merge OverlayView and ObjectOverlayHelper?

File: examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift (hunk 6 above)
Comment (khanhlvg) [reply examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift:22]: Can you add it as a sublayer like this?
https://guides.codepath.com/ios/Creating-a-Custom-Camera-View#step-11-configure-the-live-preview

File: examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift [hunk 14]
```diff
@@ -0,0 +1,358 @@
...
+      player.replaceCurrentItem(with: playerItem)
+    }
+    else {
+      playerViewController?.player = AVPlayer(playerItem: playerItem)
+    }
+    
+    playerViewController?.showsPlaybackControls = false
+    addPlayerViewControllerAsChild()
+    
+    NotificationCenter.default
```
[the remark this answers is not in the document]
Comment (khanhlvg) [reply examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift:341]: KVO for DetectorMetadata is fine because multiple classes needs to listen to it.

For this class, what about using a standard Timer to display the bounding boxes for simplicity? IIRC the original sample used this approach. We don't allow pause/resume the video player so the logic doesn't need to be tightly coupled.
- The timer ticks periodically right after the video starts
- If the timer's time exceed the video type, then clear all object bounding boxes. The lag time between when the video ends to clearing the bounding boxes will be 300ms at max, which I think is acceptable for the sake of code simplicity.

[the remark this answers is not in the document]
Comment (priankakariatyml) [reply examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift:341]: I tried it with a timer. There are other notifications we will end up adding to handle display when app enters background and foreground in this case.  When the app goes to the background, the playback is automatically paused. iOS does not resume this when the app comes back to the foreground ( can't change this behaviour).  If we use the `addPeriodicTimeObserver` on avplayer for drawing, this callback also gets paused automatically when player is paused by the OS. But if we use our own timer, we will have to register for the `app did become inactive notification `to invalidate the timer. Otherwise when you bring it back to foreground, you will see a paused player display on which boxes are being drawn periodically because the NSTimer is still active.

We can void the avplayeritem did play to end notification by using the calculation you outlined even if we retain the player's periodic time observer.

Bottom line is that either we need the player's time observer or the notification when the app enters background. 

FYI, whether we use, the player or the periodic time observer if the video playback should be resumed when the app enters foreground, we will have to register for the app did become active notification. But since this is an example app, we can choose to not handle this.

File: examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift (hunk 6 above)
Comment (priankakariatyml) [reply examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift:22]: I tried this. The frame of the added sublayer has to be updated programatically in viewWilllayoutsubviews when the device orientation changes. There is a slight glitch as the rotation and subsequent updation of the frame. I tried making this more smooth. But don't think it is possible.

https://github.com/googlesamples/mediapipe/assets/42828735/8c0d71d6-bc90-417f-b1a5-58792842b3a7

As opposed to overriding the layer subclass in a custom view subclass (Preview View). UIKit handles the rotation along with the view in this case. We don't need to update the frame anywhere.

https://github.com/googlesamples/mediapipe/assets/42828735/04bad040-743c-4dde-893d-c2f102d825a9

If this glitch during rotation and the extra frame updates are okay to include, I'll remove the PreviewView subclass and make it a sublayer. LMK

File: examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift (hunk 5 above)
Comment (priankakariatyml) [reply examples/object_detection/ios/ObjectDetector/Helpers/ObjectOverlayHelper.swift]: Fixed

File: examples/object_detection/ios/ObjectDetector/ViewContollers/ViewController.swift [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import MediaPipeTasksVision
+import UniformTypeIdentifiers
+import AVKit
+
+class ViewController: UIViewController {
```
[the remark this answers is not in the document]
Comment (priankakariatyml) [reply examples/object_detection/ios/ObjectDetector/ViewContollers/ViewController.swift:32]: Added comments

File: examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift (hunk 6 above)
Comment (khanhlvg) [reply examples/object_detection/ios/ObjectDetector/Views/PreviewView.swift:22]: I think the glitch is minor that I'd like to opt for a simplified sample app. Can you make it a sublayer?

File: examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (khanhlvg) [reply examples/object_detection/ios/ObjectDetector/ViewContollers/MediaLibraryViewController.swift]: Thanks for the thoughtful consideration. I looked at the code one more time and I think it's concise enough as it. Let's keep your current implementation.

nit: I'd suggest adding some comments to playVideo method to explain what you do there, so the developers who read the code can quickly get an idea of what you're doing in the method.

**Comment by khanhlvg** [comment]
@priankakariatyml There's a typo in the folder name.
image
